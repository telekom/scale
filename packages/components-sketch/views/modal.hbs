{{!--
Scale https://github.com/telekom/scale

Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG

This Source Code Form is subject to the terms of the Mozilla Public
License, v. 2.0. If a copy of the MPL was not distributed with this
file, You can obtain one at https://mozilla.org/MPL/2.0/.
--}}

<div data-sketch-artboard="Link" class="artboard">
  <h1 class="title--artboard">Modal</h1>

  <h2 class="title--artboard-section">Small</h2>
  <div class="agent-states--modal-desktop">
    <iframe
      frameborder="0"
      width="100%"
      height="500px"
      src="/modal-small"
    ></iframe>
  </div>

  <h2 class="title--artboard-section">Standard</h2>
  <div class="agent-states--modal-desktop">
    <iframe
      frameborder="0"
      width="100%"
      height="500px"
      src="/modal-Standard"
    ></iframe>
  </div>

  <h2 class="title--artboard-section">Large</h2>
  <div class="agent-states--modal-desktop">
    <iframe
      frameborder="0"
      width="100%"
      height="500px"
      src="/modal-large"
    ></iframe>
  </div>
</div>
