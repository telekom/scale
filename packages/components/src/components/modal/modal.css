/** 
* @prop --modal-backdrop-background: Color of the backdrop of the modal
* @prop --modal-opened-top: Top position of the modal when it is opened
* @prop --modal-background: Background color of the modal
* @prop --modal-color: Color of the text in the modal
* @prop --modal-padding: Padding of the modal
* @prop --modal-max-width: Max width of the modal
* @prop --modal-border-radius: Border radius of the modal
* @prop --modal-box-shadow: Box shadow of the modal
* @prop --modal-header-font-size: Font size of the modal header 
* @prop --modal-header-color: Color of the text of the modal header
* @prop --modal-header-margin: Margin of the modal header
* @prop --modal-header-border-bottom: Border bottom of the modal header
* @prop --modal-header-padding: Padding of the modal header
* @prop --modal-header-slot-margin: Margin of the modal header slot
* @prop --modal-header-slot-font-size: Font size of the modal header slot
* @prop --modal-header-slot-color: Color of the text of the modal header slot
* @prop --modal-header-slot-font-weight: Font weight of the modal header slot
* @prop --modal-body-padding: Padding of the modal body
* @prop --modal-body-border-bottom: Border bottom of the modal body
* @prop --modal-close-font-size: Font size of the modal close button
* @prop --modal-close-height: Height of the modal close button
* @prop --modal-close-opacity: Opacity of the modal close button
* @prop --modal-close-hover-opacity: Hover opacity of the modal close button
* @prop --modal-actions-padding: Padding of the modal actions
*/

.modal__backdrop {
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100vh;
  background: var(--modal-backdrop-background, rgba(0, 0, 0, 0.75));
  z-index: 10;
  opacity: 0;
  pointer-events: none;
}

:host([opened]) .modal__backdrop,
:host([opened]) .modal {
  opacity: 1;
  pointer-events: all;
}

:host([opened]) .modal {
  top: var(--modal-opened-top, 15vh);
}

.modal {
  box-sizing: border-box;
  overflow: hidden;
  position: fixed;
  background: var(--modal-background, white);
  color: var(--modal-color, #333);
  padding: var(--modal-padding, 1rem);
  text-align: left;
  position: fixed;
  top: 10vh;
  left: 25%;
  width: 50%;
  max-width: var(--modal-max-width, 550px);
  border-radius: var(--modal-border-radius, 3px);
  box-shadow: var(--modal-box-shadow, 0 2px 8px rgba(0, 0, 0, 0.26));
  z-index: 100;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  opacity: 0;
  pointer-events: none;
  transition: all 0.5s ease-out;
}

.modal__header {
  font-size: var(--modal-header-font-size, 0.8rem);
  color: var(--modal-header-color, #333);
  margin: var(--modal-header-margin, 0);
  border-bottom: var(--modal-header-border-bottom, 1px solid #333);
  padding: var(--modal-header-padding, 0.5rem 0);
  display: flex;
  justify-content: space-between;
}
.modal__header ::slotted(*) {
  margin: var(--modal-header-slot-margin, 0);
  font-size: var(--modal-header-slot-font-size, 1rem);
  color: var(--modal-header-slot-color, #262626);
  font-weight: var(--modal-header-slot-font-weight, bold);
}

.modal__body {
  padding: var(--modal-body-padding, 1rem 0);
  display: flex;
  justify-content: space-between;
}

.modal__close {
  font-size: var(--modal-close-font-size, 1rem);
  height: var(--modal-close-height, 16px);
  opacity: var(--modal-close-opacity, 0.5);
  cursor: pointer;
}

.modal__close:hover {
  opacity: var(--modal-close-hover-opacity, 1);
}

.modal__actions {
  border-top: var(--modal-body-border-bottom, 1px solid #ccc);
  padding: var(--modal-actions-padding, 1rem 0);
  display: flex;
  justify-content: flex-end;
}

.modal__actions t-button {
  margin: 0.25rem;
}
