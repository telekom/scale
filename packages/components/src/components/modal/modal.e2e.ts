import { newE2EPage } from '@stencil/core/testing';

describe('t-modal', () => {
  it('renders', async () => {
    const page = await newE2EPage();
    await page.setContent('<t-modal>Notification</t-modal>');
    const element = await page.find('t-modal');
    expect(element).toHaveClass('hydrated');
  });
});
