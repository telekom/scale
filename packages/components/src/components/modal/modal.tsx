import { Component, Prop, h, Method, Element } from '@stencil/core';
import { HTMLStencilElement } from '@stencil/core/internal';
import { CssClassMap } from '../../utils/utils';
import classNames from 'classnames';

@Component({
  tag: 't-modal',
  styleUrl: 'modal.css',
  shadow: true,
})
export class Modal {
  @Element() public hostElement: HTMLStencilElement;
  /** (optional) Modal class */
  @Prop() public customClass?: string = '';
  /** (optional) Modal size */
  @Prop() public size?: string = '';
  /** (optional) Modal theme */
  @Prop() public theme?: string = '';
  /** (optional) Modal variant */
  @Prop() public variant?: string = '';
  /** (required) Modal opened */
  @Prop({ reflectToAttr: true }) public opened?: boolean = false;

  private hasSlotHeader: boolean;
  private hasSlotClose: boolean;
  private hasSlotActions: boolean;

  /** Modal method: open() */
  @Method()
  public async open() {
    this.opened = true;
  }

  public closeModal = () => {
    this.opened = false;
  };

  /** Modal method: onCloseModal() */
  @Method()
  public async close() {
    this.closeModal();
  }

  public componentWillLoad() {
    this.hasSlotHeader = !!this.hostElement.querySelector('[slot="header"]');
    this.hasSlotClose = !!this.hostElement.querySelector('[slot="close"]');
    this.hasSlotActions = !!this.hostElement.querySelector(
      '[slot="modal-actions"]'
    );
  }

  public render() {
    if (!this.opened) {
      return null;
    }

    return (
      <div class={this.getCssClassMap()}>
        <div class="modal__backdrop" onClick={this.closeModal}></div>

        <div class="modal">
          {this.hasSlotHeader /* istanbul ignore next */ && (
            <div class="modal__header">
              <slot name="header" />
              <a class="modal__close" onClick={this.closeModal}>
                {this.hasSlotClose ? (
                  <div class="modal__close-icon">
                    <slot name="close" />
                  </div>
                ) : (
                  'x'
                )}
              </a>
            </div>
          )}

          <div class="modal__body">
            <slot />
          </div>

          {this.hasSlotActions /* istanbul ignore next */ && (
            <div class="modal__actions">
              <slot name="modal-actions" />
            </div>
          )}
        </div>
      </div>
    );
  }

  private getCssClassMap(): CssClassMap {
    return classNames(
      'modal',
      this.customClass && this.customClass,
      this.size && `modal--size-${this.size}`,
      this.theme && `modal--theme-${this.theme}`,
      this.variant && `modal--variant-${this.variant}`
    );
  }
}
