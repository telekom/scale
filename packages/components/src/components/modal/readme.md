# t-modal



<!-- Auto Generated Below -->


## Properties

| Property  | Attribute | Description              | Type      | Default |
| --------- | --------- | ------------------------ | --------- | ------- |
| `close`   | `close`   |                          | `string`  | `"x"`   |
| `opened`  | `opened`  |                          | `boolean` | `false` |
| `size`    | `size`    | (optional) Modal size    | `string`  | `""`    |
| `theme`   | `theme`   | (optional) Modal theme   | `string`  | `""`    |
| `variant` | `variant` | (optional) Modal variant | `string`  | `""`    |


## Methods

### `onCloseModal() => Promise<void>`



#### Returns

Type: `Promise<void>`



### `openModal() => Promise<void>`



#### Returns

Type: `Promise<void>`




## CSS Custom Properties

| Name                              | Description                                 |
| --------------------------------- | ------------------------------------------- |
| `--modal-actions-padding`         | Padding of the modal actions                |
| `--modal-backdrop-background`     | Color of the backdrop of the modal          |
| `--modal-background-color`        | Background color of the modal               |
| `--modal-body-border-bottom`      | Border bottom of the modal body             |
| `--modal-body-padding`            | Padding of the modal body                   |
| `--modal-border-radius`           | Border radius of the modal                  |
| `--modal-box-shadow`              | Box shadow of the modal                     |
| `--modal-close-font-size`         | Font size of the modal close button         |
| `--modal-close-height`            | Height of the modal close button            |
| `--modal-close-hover-opacity`     | Hover opacity of the modal close button     |
| `--modal-close-opacity`           | Opacity of the modal close button           |
| `--modal-color`                   | Color of the text in the modal              |
| `--modal-header-border-bottom`    | Border bottom of the modal header           |
| `--modal-header-color`            | Color of the text of the modal header       |
| `--modal-header-font-size`        | Font size of the modal header               |
| `--modal-header-margin`           | Margin of the modal header                  |
| `--modal-header-padding`          | Padding of the modal header                 |
| `--modal-header-slot-color`       | Color of the text of the modal header slot  |
| `--modal-header-slot-font-size`   | Font size of the modal header slot          |
| `--modal-header-slot-font-weight` | Font weight of the modal header slot        |
| `--modal-header-slot-margin`      | Margin of the modal header slot             |
| `--modal-max-width`               | Max width of the modal                      |
| `--modal-opened-top`              | Top position of the modal when it is opened |
| `--modal-padding`                 | Padding of the modal                        |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
