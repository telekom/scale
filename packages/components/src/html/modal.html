<!DOCTYPE html>
<html dir="ltr" lang="en">
  <head>
    <meta charset="utf-8" />
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=5.0"
    />
    <title>Stencil Component Starter</title>

    <script type="module" src="/build/scale-components.esm.js"></script>
    <link rel="stylesheet" href="/build/scale-components.css" />
    <style type="text/css" media="screen, print">
      html {
        margin: 0;
        padding: 0;
      }

      body {
        margin: 0;
        padding: 4rem;
      }
    </style>
  </head>

  <body>
    <h3>Modal with header</h3>
    <scale-button id="show-modal"> show modal </scale-button>
    <scale-modal id="modal-default" opened="true">
      <p>Hello. Welcome. What a pleasure it is to have you.</p>
      <scale-button slot="action"> Primary Action </scale-button>
    </scale-modal>

    <script>
      const showModal = document.querySelector('#show-modal');
      const modal = document.getElementById('modal-default');
      showModal.addEventListener('click', () => {
        modal.setAttribute('opened', 'true');
      });

      modal.addEventListener('scaleClose', () => {
        modal.removeAttribute('opened');
      });

      modal.addEventListener('scaleBeforeClose', (event) => {
        // event.preventDefault()
      });
    </script>
  </body>
</html>
