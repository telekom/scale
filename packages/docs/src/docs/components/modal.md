---
title: Modal
description: Modal element
url: /docs/modal
contributors:
  - SylviaGonschior
---

# Modal

Modal component

selector: `t-modal`

## Modal with header

```javascript
 <t-button id="show-modal">
      show modal
  </t-button>
    <t-modal id="modal" opened="false">
      <span slot="header">Header content</span> content in the modal slot
    </t-modal>
  <script>
      const showModal = document.querySelector('#show-modal');
      const modal = document.getElementById('modal');
      showModal.addEventListener('click', () => !modal.opened && modal.open());
  </script>
```

<div class="demo-container">
  <t-button id="show-defaultmodal">
      show modal
    </t-button>
 <t-modal id='modal-default' opened='false'>
 <span slot='header'>Header content</span> content in the modal slot
 </t-modal>
  <script>
      const showModalDefault = document.querySelector('#show-defaultmodal');
      const modalDefault = document.getElementById('modal-default');
      showModalDefault.addEventListener('click', () => !modalDefault.opened && modalDefault.open());
    </script>

</div>

## Modal without header

```javascript
  <t-button id="show-modal-without-header">
      show modal without header
    </t-button>
    <t-modal id="modal-without-header" opened="false">content</t-modal>
    <script>
      const showModalWithoutHeader = document.querySelector(
        '#show-modal-without-header'
      );
      const modalWithoutHeader = document.getElementById(
        'modal-without-header'
      );
      showModalWithoutHeader.addEventListener(
        'click',
        () => !modalWithoutHeader.opened && modalWithoutHeader.open()
      );
    </script>
```

<div class="demo-container">
  <t-button id="show-modal-without">
      show modal without header
    </t-button>
    <t-modal id="modal-without" opened="false">content</t-modal>
    <script>
      const showModalWithout = document.querySelector(
        '#show-modal-without'
      );
      const modalWithout = document.getElementById(
        'modal-without'
      );
      showModalWithout.addEventListener(
        'click',
        () => !modalWithout.opened && modalWithout.open()
      );
    </script>
</div>

## Modal with header and buttons

```javascript
<t-button id="show-modal-with-header-and-buttons">
      show modal with header and buttons
    </t-button>
    <t-modal id="modal-with-header-and-buttons" opened="false">
      <span slot="header">Header</span> content of the modal with header and
      buttons
      <span slot="close">Close</span>
      <span slot="modal-actions">
        <t-button id="cancel-btn">cancel</t-button>
        <t-button id="submit-btn">submit</t-button>
      </span>
    </t-modal>

    <script>
      const showModalWithHeaderAndButtons = document.querySelector(
        '#show-modal-with-header-and-buttons'
      );
      const modalWithHeaderAndButtons = document.getElementById(
        'modal-with-header-and-buttons'
      );
      const cancelModal = document.getElementById('cancel-btn');

      showModalWithHeaderAndButtons.addEventListener(
        'click',
        () =>
          !modalWithHeaderAndButtons.opened && modalWithHeaderAndButtons.open()
      );

      cancelModal.addEventListener('click', () =>
        modalWithHeaderAndButtons.close()
      );
    </script>
```

<div class="demo-container">
  <t-button id="show-modal-header-and-buttons">
      show modal with header and buttons
    </t-button>
    <t-modal id="modal-header-and-buttons" opened="false">
      <span slot="header">Header</span> content of the modal with header and
      buttons
      <span slot="close">Close</span>
      <span slot="modal-actions">
        <t-button id="cancel-button">cancel</t-button>
        <t-button id="submit-btn">submit</t-button>
      </span>
    </t-modal>
    <script>
      const showModalHeaderAndButtons = document.querySelector(
        '#show-modal-header-and-buttons'
      );
      const modalHeaderAndButtons = document.getElementById(
        'modal-header-and-buttons'
      );
      const cancelModalBtn = document.getElementById('cancel-button');
      showModalHeaderAndButtons.addEventListener(
        'click',
        () =>
          !modalHeaderAndButtons.opened && modalHeaderAndButtons.open()
      );
      cancelModalBtn.addEventListener('click', () =>
        modalHeaderAndButtons.close()
      );
    </script>
</div>
