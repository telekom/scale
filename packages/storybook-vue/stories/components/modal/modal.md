<div style="display: inline-flex; align-items: center; justify-content: space-between; width: 100%;">
    <h1>Modal</h1>
    <img src="assets/aa.png" alt="Accessible AA" />
</div>

Modal dialogs require immediate attention and interrupt the user's workflow. They appear whenever the user needs information or has to make a decision to complete a task.

## General

A modal dialog appears as an overlay. The main content of the page is disabled until the user interacts with the modal.

### When to use

Use a modal if:

- you need to provide warnings and urgent information about the user's current task (e.g., system errors or the consequence of a user's action),
- the system requires critical information from the user to execute the current process,
- particular information from the user would make a process easier for them to continue, or
- the user needs to confirm a decision to prevent or correct critical errors.

### When not to use

Don't use a modal for:

- information or actions that are not related to the current user flow (this might prevent users from completing their task) or
- complex decision making that requires information the modal doesn't provide (this might force users to switch between the modal and main page, which would critically affect the user experience).

## Elements

![Image Name](assets/3_components/modal/Modal_Basic.png)

### Title (1)

A title is mandatory and should be short and concise.

### Content (2)

Different kinds of content are possible. Once the modal reaches its maximum display height, the content starts scrolling.

#### Action area (3)

The action area can contain one primary button and up to two secondary buttons. The primary button should represent a positive action (confirm, save); the secondary buttons represent negative or less important actions (delete, cancel).

### Close Icon (4)

The close icon is mandatory.

## Scrolling

If the content exceeds the space provided by the modal, the content becomes scrollable.
To frame the scrollable area, an overlay highlights the action bar. A divider appears below the header.

![Image Name](assets/3_components/modal/Modal_Scrolling.png)

## Sizes

### Desktop

For desktops, modals come in sizes S, L, or XL and span 6, 8, or 12 columns, respectively. Select the modal size that best fits the amount of content for your use case.

![Image Name](assets/3_components/modal/desktop_12columns.png)
![Image Name](assets/3_components/modal/desktop_8columns.png)
![Image Name](assets/3_components/modal/desktop_6columns.png)

### Mobile

On mobile, all modals have the same size: full width (4 columns).

![Image Name](assets/3_components/modal/mobile_4columns.png)

## Best practices

- Clearly describe the confirmed action and explain the potential consequences.
- Both title and button should indicate which action will occur.
- Never use a primary button if the action is destructive or irreversible.
