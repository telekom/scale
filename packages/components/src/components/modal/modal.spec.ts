import { newSpecPage } from '@stencil/core/testing';
import { Modal } from './modal';

describe('Modal', () => {
  let element;
  beforeEach(async () => {
    element = new Modal();
  });

  const components = [Modal];

  it('should match snapshot', async () => {
    const page = await newSpecPage({
      components: [Modal],
      html: `<t-modal>Label</t-modal>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('should match snapshot with header slot', async () => {
    const page = await newSpecPage({
      components: [Modal],
      html: `
			<t-modal>
				<span slot="header">Header content</span>
				A title
			</t-modal>
			`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('should match snapshot with actions slot', async () => {
    const page = await newSpecPage({
      components: [Modal],
      html: `
			<t-modal>
				<span slot="modal-actions">Action buttons</span>
				Content
			</t-modal>
			`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('should match snapshot when opened', async () => {
    const page = await newSpecPage({
      components,
      html: `<t-modal opened=true >Label</t-modal>`,
    });
    expect(page.root.shadowRoot).toBeTruthy();
    expect(page.root).toMatchSnapshot();
  });

  it('should handle size css class', () => {
    element.size = 'small';
    expect(element.getCssClassMap()).toContain('modal--size-small');
  });

  it('should handle theme css class', () => {
    element.theme = 'default';
    expect(element.getCssClassMap()).toContain('modal--theme-default');
  });

  it('should handle variant css class', () => {
    element.variant = 'primary';
    expect(element.getCssClassMap()).toContain('modal--variant-primary');
  });

  it('should open the modal', () => {
    expect(element.opened).toBe(false);
    element.open();
    expect(element.opened).toBe(true);
  });

  it('should close the modal', () => {
    element.close();
    expect(element.opened).toBe(false);
  });

  it('should close the modal', () => {
    element.closeModal();
    expect(element.opened).toBe(false);
  });

  it('should not open the modal/ should not render, if the modal is already opened', () => {
    element.opened = true;
    expect(element.root).toBeFalsy();
  });
});
