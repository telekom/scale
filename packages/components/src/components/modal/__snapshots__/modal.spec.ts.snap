// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should match snapshot 1`] = `
<t-modal>
  <mock:shadow-root></mock:shadow-root>
  Label
</t-modal>
`;

exports[`Modal should match snapshot when opened 1`] = `
<t-modal opened="">
  <mock:shadow-root>
    <div class="modal">
      <div class="modal__backdrop"></div>
      <div class="modal">
        <div class="modal__body">
          <slot></slot>
        </div>
      </div>
    </div>
  </mock:shadow-root>
  Label
</t-modal>
`;

exports[`Modal should match snapshot with actions slot 1`] = `
<t-modal>
  <mock:shadow-root></mock:shadow-root>
  <span slot="modal-actions">
    Action buttons
  </span>
  Content
</t-modal>
`;

exports[`Modal should match snapshot with header slot 1`] = `
<t-modal>
  <mock:shadow-root></mock:shadow-root>
  <span slot="header">
    Header content
  </span>
  A title
</t-modal>
`;
